### DISCUSSION TITLE
Export more of the internal Keystone types to allow decoupling schema definition from behaviors

### DISCUSSION THREAD

**Discussion by remygendron** [keystonejs/keystone#7298] [Feature requests]
Hi there!

To keep my schema file clean, I want to extract things like virtual field and hooks definitions into separate files (eg: listname-validation.ts, listname-virtualfields.ts). Schema parts that come with a builder like virtual() are easy, as when you import this helper function you get typings to satisfy TS. But with hooks, it ain’t that easy. I can’t import ValidateInputHook<ListTypeInfo> to type my validator function in a separate file.

What would be the best approach to do that? Big thank you.

**Comment by elliott-w** [comment]
For anybody who comes across this, I'm doing something like this for hooks:

In `src/utils/slug.ts`:
```ts
import { TypeInfo } from '.keystone/types'
import { ListHooks } from '@keystone-6/core/types'

export const resolveSlugService: NonNullable<
  ListHooks<TypeInfo['lists']['Service']>['resolveInput']
> = ({ resolvedData }) => {
  // Do stuff here
  return resolvedData
}
```

In `src/lists/Service.ts`:
```ts
import { Lists } from '.keystone/types'
import { list } from '@keystone-6/core'
import { text } from '@keystone-6/core/fields'
import { resolveSlugService } from '../utils/slug'

const lists: Lists = {
  Service: list({
    fields: {
      title: text({
        validation: {
          isRequired: true,
        },
      }),
      slug: text({
        isIndexed: 'unique',
        ui: {
          description: 'Leave blank to auto generate',
        },
      }),
    },
    hooks: {
      resolveInput: args => {
        const resolvedData = resolveSlugService(args)
        // Some generic function here to pipe resolvedData through different functions would be nice
        return resolvedData
      },
    },
  }),
}

export default lists
```

One thing I found really hard/impossible is to make a typed `resolveSlug()` function that works for different list types. Especially if the slug is not always made from the same field name (e.g. title).
